metaflow/metaflow_system_logger.py

```diff
@@ -0,0 +1,101 @@
+import sys
+from typing import Dict, Any, Optional, Union
+
+
+class SystemLogger(object):
+    def __init__(self):
+        self._flow = None
+        self._environment = None
+        self._logger = None
+        self._flow_name = None
+
+    def __del__(self):
+        if self.flow_name == "not_a_real_flow":
+            self.logger.terminate()
+
+    @property
+    def environment(self) -> Optional["metaflow_environment.MetaflowEnvironment"]:
+        from .plugins import ENVIRONMENTS
+        from .metaflow_config import DEFAULT_ENVIRONMENT
+        from .metaflow_environment import MetaflowEnvironment
+
+        if self._environment is None:
+            self._environment = [
+                e
+                for e in ENVIRONMENTS + [MetaflowEnvironment]
+                if e.TYPE == DEFAULT_ENVIRONMENT
+            ][0](self.flow)
+        return self._environment
+
+    @property
+    def flow(
+        self,
+    ) -> Optional[Union["metaflow.flowspec.FlowSpec", "metaflow.sidecar.DummyFlow"]]:
+        from metaflow.sidecar import DummyFlow
+
+        if self._flow is None:
+            self._flow = DummyFlow()
+            self._flow_name = self._flow.name
+        return self._flow
+
+    @property
+    def flow_name(self) -> Optional[str]:
+        return self._flow_name
+
+    @property
+    def logger(self) -> Optional["metaflow.event_logger.NullEventLogger"]:
+        from .plugins import LOGGING_SIDECARS
+        from .metaflow_config import DEFAULT_EVENT_LOGGER
+
+        if self._logger is None:
+            self._logger = LOGGING_SIDECARS[DEFAULT_EVENT_LOGGER](
+                flow=self.flow, env=self.environment
+            )
+            self._debug("Started logger outside of a flow")
```

Comment:
Is this always true (or does this distinction even matter)?

Reply:
It is not always true. `self._logger` would be `None` only when the logger is being accessed outside a flow. Otherwise we set `self._logger` to the event logger in `cli.py`.